uast/iter.go

```diff
@@ -0,0 +1,97 @@
+package uast
+
+import (
+	"sort"
+
+	"gopkg.in/bblfsh/sdk.v2/uast/nodes"
+)
+
+// NewPositionalIterator creates a new iterator that enumerates all object nodes, sorting them by positions in the source file.
+// Nodes with no positions will be enumerated last.
+func NewPositionalIterator(root nodes.External) nodes.Iterator {
+	return &posIterator{root: root}
+}
+
+type posIterator struct {
+	root  nodes.External
+	nodes []nodes.External
+}
+
+type nodesByPos struct {
+	nodes []nodes.External
+	pos   []Position
+}
+
+func (arr nodesByPos) Len() int {
+	return len(arr.nodes)
+}
+
+func (arr nodesByPos) Less(i, j int) bool {
+	return arr.pos[i].Less(arr.pos[j])
+}
+
+func (arr nodesByPos) Swap(i, j int) {
+	arr.nodes[i], arr.nodes[j] = arr.nodes[j], arr.nodes[i]
+	arr.pos[i], arr.pos[j] = arr.pos[j], arr.pos[i]
+}
+
+func (it *posIterator) sort() {
+	// in general, we cannot expect that parent node's positional info will include all the children
+	// because of this we are forced to enumerate all nodes and sort them by positions on the first call to Next
+	var plist []Position
+	noPos := func() {
+		plist = append(plist, Position{})
+	}
+	posType := TypeOf(Positions{})
+	nodes.WalkPreOrderExt(it.root, func(n nodes.External) bool {
```

Comment:
Can't this use the existing PreOrder iterator?

Reply:
It will be impossible to skip nodes. This iterator also skips positional nodes.